File: frontend/src/stores/portCargo.js
import { defineStore } from 'pinia'
import { portCargoApi } from '../api/port'

export const usePortCargoStore = defineStore('portCargo', {
  state: () => ({
    containers: [],
    total: 0,
    currentContainer: null,
    traceEvents: [],
    statistics: null,
    loading: false,
  }),

  actions: {
    async searchContainers(params) {
      this.loading = true
      try {
        const { data } = await portCargoApi.searchContainers(params)
        this.containers = data.items
        this.total = data.total
      } finally {
        this.loading = false
      }
    },

    async getContainer(id) {
      const { data } = await portCargoApi.getContainer(id)
      this.currentContainer = data
    },

    async getTrace(id) {
      const { data } = await portCargoApi.getContainerTrace(id)
      this.traceEvents = data.items
    },

    async fetchStatistics() {
      const { data } = await portCargoApi.getStatistics()
      this.statistics = data
    },
  },
})

File: backend/app/api/port_cargo.py
from fastapi import APIRouter, Depends, Query, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from app.database import get_db
from app.services import port_cargo_service
from app.schemas.port_cargo import ContainerCreate, ContainerEventCreate

router = APIRouter(prefix="/port/cargo", tags=["port-cargo"])


@router.get("/containers")
async def search_containers(
    keyword: str = Query(None),
    vessel: str = Query(None),
    status: str = Query(None),
    start_date: str = Query(None),
    end_date: str = Query(None),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
):
    return await port_cargo_service.search_containers(
        db, keyword, vessel, status, start_date, end_date, page, page_size
    )


@router.post("/containers")
async def create_container(data: ContainerCreate, db: AsyncSession = Depends(get_db)):
    return await port_cargo_service.create_container(db, data.model_dump())


@router.get("/containers/{container_id}")
async def get_container(container_id: int, db: AsyncSession = Depends(get_db)):
    container = await port_cargo_service.get_container(db, container_id)
    if not container:
        raise HTTPException(status_code=404, detail="Container not found")
    return container


@router.get("/containers/{container_id}/trace")
async def get_trace(container_id: int, db: AsyncSession = Depends(get_db)):
    events = await port_cargo_service.get_container_trace(db, container_id)
    return {"items": events}


@router.post("/containers/{container_id}/events")
async def add_event(container_id: int, data: ContainerEventCreate, db: AsyncSession = Depends(get_db)):
    container = await port_cargo_service.get_container(db, container_id)
    if not container:
        raise HTTPException(status_code=404, detail="Container not found")
    return await port_cargo_service.add_event(db, container_id, data.model_dump())


@router.get("/statistics")
async def get_statistics(db: AsyncSession = Depends(get_db)):
    return await port_cargo_service.get_statistics(db)
